kahypar/partition/bin_packing/bin_packing.h

```diff
@@ -0,0 +1,458 @@
+            bin_counts[part_idx]++;
+            if (bin_counts[part_idx] >= num_bins_per_part[part_idx]) {
+              partition_packer.lockBin(part);
+            }
+
+            mapping.setPart(bin, part);
+          }
+        }
+
+        // collect part weights
+        std::vector<HypernodeWeight> part_weights;
+        part_weights.reserve(num_parts);
+        for (PartitionID i = 0; i < num_parts; ++i) {
+          part_weights.push_back(partition_packer.binWeight(i) + max_allowed_part_weights[i] - max_part_weight);
+        }
+
+        return {std::move(mapping), std::move(part_weights)};
+      }
+
+    private:
+      BPAlgorithm _alg;
+      PartitionMapping _bins_to_parts;
+  };
+
+  // Returns the current hypernodes sorted in descending order of weight.
+  static inline std::vector<HypernodeID> extractNodesWithDescendingWeight(const Hypergraph& hg) {
+    std::vector<HypernodeID> nodes;
+    nodes.reserve(hg.currentNumNodes());
+
+    for (const HypernodeID& hn : hg.nodes()) {
+      nodes.push_back(hn);
+    }
+    ASSERT(hg.currentNumNodes() == nodes.size());
+
+    std::sort(nodes.begin(), nodes.end(), [&hg](HypernodeID a, HypernodeID b) {
+      return hg.nodeWeight(a) > hg.nodeWeight(b);
+    });
+
+    return nodes;
+  }
+
+  // Preassigns the fixed vertices to the packer with a first fit packing.
+  template< class BPAlgorithm >
+  static inline void preassignFixedVertices(const Hypergraph& hg, const std::vector<HypernodeID>& nodes, std::vector<PartitionID>& parts,
+                                            TwoLevelPacker<BPAlgorithm>& packer, PartitionID k, PartitionID rb_range_k) {
+    const HypernodeWeight avg_bin_weight = (hg.totalWeight() + rb_range_k - 1) / rb_range_k;
+    const PartitionID bins_per_part = (rb_range_k + k - 1) / k;
+
+    for (size_t i = 0; i < nodes.size(); ++i) {
+      const HypernodeID hn = nodes[i];
+
+      if (hg.isFixedVertex(hn)) {
+        const HypernodeWeight weight = hg.nodeWeight(hn);
+        const PartitionID part_id = hg.fixedVertexPartID(hn);
+        const PartitionID start_index = part_id * bins_per_part;
+        PartitionID assigned_bin = start_index;
+
+        if (bins_per_part > 1) {
+          for (PartitionID i = start_index; i < std::min(start_index + bins_per_part, rb_range_k); ++i) {
+            HypernodeWeight current_bin_weight = packer.binWeight(i);
+
+            // The node is assigned to the first fitting bin or, if none fits, the smallest bin.
+            if (current_bin_weight + weight <= avg_bin_weight) {
+              assigned_bin = i;
+              break;
+            } else if (current_bin_weight < packer.binWeight(assigned_bin)) {
+              assigned_bin = i;
+            }
+          }
+        }
+
+        packer.addFixedVertex(assigned_bin, part_id, weight);
+        parts[i] = assigned_bin;
+      }
+    }
+
+  }
+
+
+  // segment tree that is used to ensure the balance of a prepacking
```

Comment:
balance_max, balance_base, BalanceSegTree and getMaxPartIndex should be in a anonymous namespace
namespace {
 ....
}